### PR TITLE
Feature 4

### PR DESCRIPTION
**Pull request by a99jafari**

### CODE REVIEW

File: src/components/App.jsx [hunk 1]
```diff
@@ -47,9 +47,25 @@ const App = () => {
...
     );
 
+    const handleCompleted = useCallback(
+        (targetTodo) => {
+            let filter = [...todos];
+            const index = todos.findIndex((todo) => {
+                return todo === targetTodo;
+            });
+            filter[index].completed = !filter[index].completed;
+            setTodos(filter);
```
Comment (sassanh) [inline src/components/App.jsx:57]: We rarely use `findIndex` (only when index really matters, here index doesn't matter, the particular item matters)
So you can implement this function like this:
```javascript
setTodos(todos.map((todoItem) => todoItem === targetTodo ? {...todoItem, completed: !todoItem.completed} : todoItem));
```

File: src/context.js [hunk 2]
```diff
@@ -6,6 +6,7 @@ const todoContext = createContext({
     handleChange: () => {},
     handleAdd: () => {},
     handleRemove: () => {},
+    handleCompleted: () => {},
```
Comment (sassanh) [inline src/context.js:9]: We agreed to remove this file altogether, please check your diff here after you create the pull request to make sure it is exactly as expected.

File: src/components/Todo.jsx [hunk 3]
```diff
@@ -1,13 +1,23 @@
 import React, { memo } from "react";
 
-const Todo = ({ todo, onRemove }) => {
+const Todo = ({ todo, onRemove, onComplete }) => {
     const handleRemove = () => {
```
Comment (sassanh) [inline src/components/Todo.jsx:4]: Needs `useCallback`

File: src/components/Todo.jsx [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
         onRemove(todo);
     };
 
+    const handleComplete = () => {
```
Comment (sassanh) [inline src/components/Todo.jsx:8]: Needs `useCallback`
(As we communicated earlier: Every callback should be wrapped inside a `useCallback` with appropriate dependencies)

**Review by sassanh — CHANGES_REQUESTED** [review]

File: src/components/App.jsx (hunk 1 above)
Comment (a99jafari) [reply src/components/App.jsx:57]: Thank you. I changed it.

File: src/context.js (hunk 2 above)
Comment (a99jafari) [reply src/context.js:9]: I delete this file.

File: src/components/Todo.jsx (hunk 4 above)
Comment (a99jafari) [reply src/components/Todo.jsx:8]: Sorry Sassan,
I don't know how to use of "useCallBack" in this case. Can you guide me more?

Comment (sassanh) [reply src/components/Todo.jsx:8]: Like this:
```
const handleComplete = useCallback(() => {
  onComplete(todo);
}, [onComplete, todo]);
```
Let me know what was confusing you?

**Comment by a99jafari** [comment]
Thank you Sassan,
I thought we can only pass states as dependencies to useCallback. I think I
did not understand the difference between useMemo and useCallback exactly
and I have to search and learn more about them.

Do we need to use 'useCallback' for 'handleRemove' in Todo.jsx?

**Comment by sassanh** [comment]
@a99jafari Yes we do, note that the functionality of `useCallback` and `useMemo` are the same (if used correctly) `useCallback` is there only because memoizing functions doesn't need an extra function definition because they are already functions and their body doesn't run each time your function component renders. Take your time to read about it.

About things that can be used in the dependencies of `useCallback`, `useEffect` and `useMemo` they can be return values of `useState` (the first one, not the setter), return values of `useCallback`, return values of `useMemo` and also props (later when you start using mobs we will add observable variables and computed variables to this list too) the `current` property of the return value of `useRef` specially **cannot** be used as dependency.

**Review by sassanh — APPROVED** [review]
You can consider removing `console.log` expressions, looks good to merge.

**Pull request merged**